esphome/components/madoka/madoka.cpp

```diff
@@ -0,0 +1,413 @@
+  }
+  uint8_t chunk_id = chk[0];
+  message stripped(chk.begin() + 1, chk.end());
+  if (chunk_id == 0 && validate_buffer(stripped)) {
+    this->parse_cb_(stripped);
+    return;
+  }
+  if (this->pending_chunks_.count(chunk_id)) {
+    ESP_LOGE(TAG, "Another packet with the same chunk ID is already in the buffer.");
+    ESP_LOGD(TAG, "Chunk ID: %d.", chunk_id);
+    return;
+  }
+  this->pending_chunks_[chunk_id] = chk;
+
+  if (this->pending_chunks_.size() != this->pending_chunks_.rbegin()->first + 1) {
+    ESP_LOGW(TAG, "Buffer is missing packets");
+    return;
+  }
+
+  message msg;
+  int lim = this->pending_chunks_.size();
+  for (int i = 0; i < lim; i++) {
+    msg.insert(msg.end(), this->pending_chunks_[i].begin() + 1, this->pending_chunks_[i].end());
+  }
+  if (validate_buffer(msg)) {
+    this->pending_chunks_.clear();
+    this->parse_cb_(msg);
+  }
+}
+
+std::vector<chunk> Madoka::split_payload_(message msg) {
+  std::vector<chunk> result;
+  size_t len = msg.size();
+  result.push_back(chunk({0x00, (uint8_t) (len + 1)}));
+  result[0].insert(result[0].end(), msg.begin(), min(msg.begin() + (MAX_CHUNK_SIZE - 2), msg.end()));
+  int i = 0;
+  for (i = 1; i < len / (MAX_CHUNK_SIZE - 1); i++) {  // from second to second-last
+    result.emplace_back(msg.begin() + ((MAX_CHUNK_SIZE - 1) * i - 1),
+                        msg.begin() + ((MAX_CHUNK_SIZE - 1) * (i + 1) - 1));
+  }
+  if (len > 18) {
+    i++;
+    result.emplace_back(msg.begin() + ((MAX_CHUNK_SIZE - 1) * i), msg.end());
+  }
+  return result;
+}
+
+message Madoka::prepare_message_(uint16_t cmd, message args) {
+  message result({0x00, (uint8_t) ((cmd >> 8) & 0xFF), (uint8_t) (cmd & 0xFF)});
+  result.insert(result.end(), args.begin(), args.end());
+  return result;
+}
+
+void Madoka::query_(uint16_t cmd, message args, int t_d) {
+  message payload = this->prepare_message_(cmd, std::move(args));
+
+  if (this->node_state != espbt::ClientState::ESTABLISHED) {
+    return;
+  }
+  std::vector<chunk> chunks = this->split_payload_(payload);
+
+  for (auto chk : chunks) {
+    esp_err_t status;
+    for (int j = 0; j < BLE_SEND_MAX_RETRIES; j++) {
+      status = esp_ble_gattc_write_char(this->parent_->get_gattc_if(), this->parent_->get_conn_id(), this->wwr_handle_,
+                                        chk.size(), &chk[0], ESP_GATT_WRITE_TYPE_NO_RSP, ESP_GATT_AUTH_REQ_NONE);
+      if (!status) {
+        break;
+      }
+      ESP_LOGD(TAG, "[%s] esp_ble_gattc_write_char failed (%d of %d), status=%d", this->parent_->address_str().c_str(),
+               j + 1, BLE_SEND_MAX_RETRIES, status);
+    }
+    if (status) {
+      ESP_LOGE(TAG, "[%s] Command could not be sent, last status=%d", this->parent_->address_str().c_str(), status);
+      return;
+    }
+  }
+  delay(t_d);
+}
+
+void Madoka::parse_cb_(message msg) {
+  uint16_t function_id = msg[2] << 8 | msg[3];
+  uint8_t i = 4;
+  uint8_t message_size = msg.size();
+
+  switch (function_id) {
+    case 0x0020:
```

Comment:
Named static consts please

```suggestion
    case 0x0020:
```